File: Python/0004.py
from typing import List

class Solution:
    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:
        # Ensure nums1 is the smaller array to optimize the binary search
        if len(nums1) > len(nums2):
            nums1, nums2 = nums2, nums1
        
        m, n = len(nums1), len(nums2)
        low, high = 0, m
        
        # Perform binary search on the smaller array nums1
        while low <= high:
            partition1 = (low + high) // 2  # Calculate partition for nums1
            partition2 = (m + n + 1) // 2 - partition1  # Calculate partition for nums2
            
            # Determine the elements around partitions for both arrays
            maxLeft1 = float('-inf') if partition1 == 0 else nums1[partition1 - 1]
            minRight1 = float('inf') if partition1 == m else nums1[partition1]
            
            maxLeft2 = float('-inf') if partition2 == 0 else nums2[partition2 - 1]
            minRight2 = float('inf') if partition2 == n else nums2[partition2]
            
            # Check if partitions are correct for median condition
            if maxLeft1 <= minRight2 and maxLeft2 <= minRight1:
                # If total length is even, return average of middle elements
                if (m + n) % 2 == 0:
                    return (max(maxLeft1, maxLeft2) + min(minRight1, minRight2)) / 2
                else:
                    # If total length is odd, return the larger of the two middle elements
                    return max(maxLeft1, maxLeft2)
            elif maxLeft1 > minRight2:
                # Adjust partition1 to the left if maxLeft1 is too large
                high = partition1 - 1
            else:
                # Adjust partition1 to the right if maxLeft2 is too large
                low = partition1 + 1
        
        # If input arrays are not sorted (though they should be), raise an error
        raise ValueError("Input arrays are not sorted")

File: Rust/0004.rs
impl Solution {
    pub fn find_median_sorted_arrays(nums1: Vec<i32>, nums2: Vec<i32>) -> f64 {
        let (nums1, nums2) = if nums1.len() > nums2.len() {
            (nums2, nums1) // Ensure nums1 is the smaller array
        } else {
            (nums1, nums2)
        };

        let m = nums1.len();
        let n = nums2.len();

        let mut imin = 0;
        let mut imax = m;
        let half_len = (m + n + 1) / 2;

        while imin <= imax {
            let i = (imin + imax) / 2;
            let j = half_len - i;

            let max_left1 = if i == 0 { i32::MIN } else { nums1[i - 1] };
            let min_right1 = if i == m { i32::MAX } else { nums1[i] };

            let max_left2 = if j == 0 { i32::MIN } else { nums2[j - 1] };
            let min_right2 = if j == n { i32::MAX } else { nums2[j] };

            if max_left1 <= min_right2 && max_left2 <= min_right1 {
                if (m + n) % 2 == 0 {
                    return (f64::max(max_left1 as f64, max_left2 as f64) + f64::min(min_right1 as f64, min_right2 as f64)) / 2.0;
                } else {
                    return f64::max(max_left1 as f64, max_left2 as f64);
                }
            } else if max_left1 > min_right2 {
                imax = i - 1;
            } else {
                imin = i + 1;
            }
        }

        panic!("Input arrays are not sorted properly.");
    }
}
